lib/service/kubernetes.go

```diff
@@ -0,0 +1,231 @@
+		}
+
+		err := process.initKubernetesService(log, conn)
+		if err != nil {
+			warnOnErr(conn.Close())
+			return trace.Wrap(err)
+		}
+		return nil
+	})
+}
+
+func (process *TeleportProcess) initKubernetesService(log *logrus.Entry, conn *Connector) error {
+	// clean up unused descriptors passed for proxy, but not used by it
+	defer func() {
+		if err := process.closeImportedDescriptors(teleport.ComponentKube); err != nil {
+			log.WithError(err).Warn("Failed closing imported file descriptors")
+		}
+	}()
+	cfg := process.Config
+
+	// Create a caching auth client.
+	accessPoint, err := process.newLocalCache(conn.Client, cache.ForKubernetes, []string{teleport.ComponentKube})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// This service can run in 2 modes:
+	// 1. Reachable (by the proxy) - registers with auth server directly and
+	//    creates a local listener to accept proxy conns.
+	// 2. Not reachable ("IoT mode") - creates a reverse tunnel to a proxy and
+	//    handles registration and incoming connections through that.
+	//
+	// The listener exposes incoming connections over either mode.
+	var listener net.Listener
+	var agentPool *reversetunnel.AgentPool
+	switch {
+	// Filter out cases where both listen_addr and tunnel are set or both are
+	// not set.
+	case conn.UseTunnel() && !cfg.Kube.ListenAddr.IsEmpty():
+		return trace.BadParameter("either set kubernetes_service.listen_addr if this process can be reached from a teleport proxy or point teleport.auth_servers to a proxy to dial out, but don't set both")
+	case !conn.UseTunnel() && cfg.Kube.ListenAddr.IsEmpty():
+		return trace.BadParameter("set kubernetes_service.listen_addr if this process can be reached from a teleport proxy or point teleport.auth_servers to a proxy to dial out")
+
+	// Start a local listener and let proxies dial in.
+	case !conn.UseTunnel() && !cfg.Kube.ListenAddr.IsEmpty():
+		log.Debugf("Turning on Kubernetes service listening address.")
+		listener, err = process.importOrCreateListener(listenerKube, cfg.Kube.ListenAddr.Addr)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+	// Dialed out to a proxy, start servicing the reverse tunnel as a listener.
+	case conn.UseTunnel() && cfg.Kube.ListenAddr.IsEmpty():
+		// create an adapter, from reversetunnel.ServerHandler to net.Listener.
+		shtl := reversetunnel.NewServerHandlerToListener()
+		listener = shtl
+		agentPool, err = reversetunnel.NewAgentPool(
+			process.ExitContext(),
+			reversetunnel.AgentPoolConfig{
+				Component:   teleport.ComponentKube,
+				HostUUID:    conn.ServerIdentity.ID.HostUUID,
+				ProxyAddr:   conn.TunnelProxy(),
+				Client:      conn.Client,
+				AccessPoint: conn.Client,
+				HostSigner:  conn.ServerIdentity.KeySigner,
+				Cluster:     conn.ServerIdentity.Cert.Extensions[utils.CertExtensionAuthority],
+				Server:      shtl,
+			})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		err = agentPool.Start()
```

Comment:
In case there're any errors further down this function, the agentPool will continue running - not sure if there's anything it needs to clean up on process shutdown.

Reply:
good call.
added defers to clean up `agentPool`, `listener` and `kubeServer` if this function fails